### PR TITLE
PR 2130 & PR 2233

### PR DESCRIPTION
**Pull request by Nihadness**
Integration of:
- https://github.com/bikalims/bika.lims/pull/2130
- https://github.com/bikalims/bika.lims/pull/2233

**Current behavior before PR**

1. Infinite recursion if no Address details were found on the Client or Contact.
2. Selected items in listing tables were processed in random order. If a use selected ARs from the list and clicked the copy button, the ARs were listed in random order in the AR Add form (and of course created in this random order).

**Desired behavior after PR is merged**

1. Rendering succeeds w/o Address Details.
2. Selected items in listing views keep their order and copied ARs are processed in reversed order.

### CODE REVIEW

File: bika/lims/browser/analysisrequest/publish.py [hunk 1]
```diff
@@ -1267,7 +1277,7 @@ def _client_data(self, ar):
             data['phone'] = to_utf8(client.getPhone())
             data['fax'] = to_utf8(client.getFax())
 
-            data['address'] = to_utf8(self._client_address(client))
+            data['address'] = to_utf8(self._client_address(ar))
```
[the remark this answers is not in the document]
Comment (xispa) [reply bika/lims/browser/analysisrequest/publish.py]: My bad, realize this function is from inside `AnalysisRequestDigester`. Altough its signature is the same as the function in `AnalysisRequestPublishView`, they behave differently, returning different type of data. Please, make sure that there are no inconsistences with the data generated (and later used for the report creation)  by both `AnalysisRequestDigester` and `AnalysisRequestPublishView`

**Comment by ramonski** [comment]
Ok, I see and I think you are right @xispa – the `reverse` should be removed

File: bika/lims/browser/analysisrequest/publish.py [hunk 2]
```diff
@@ -1549,3 +1515,35 @@ def EndRequestHandler(event):
...
+        # Data from the first contact
+        contact = ar.getContact()
+        if contact and contact.getBillingAddress():
+            client_address = contact.getBillingAddress()
+        elif contact and contact.getPhysicalAddress():
+            client_address = contact.getPhysicalAddress()
+    return _format_address(client_address)
+
+
+def _format_address(address):
```
Comment (ramonski) [inline bika/lims/browser/analysisrequest/publish.py]: What is the reason of having this function outside of the class?

**Comment by Nihadness** [comment]
@ramonski , just because it was in both of those classes, and actually was added to ```AnalysisRequestPublishView``` class recently (during integration I guess). Just in case someone needs it, I thought, instead of repeating it in ```AnalysisRequestDigester```, it is better to keep out of the class and use from there.

**Comment by ramonski** [comment]
Thanks for the explanation @Nihadness. I completely agree then, because it supports the DRY (don't repeat yourself) approach – good work!

**Pull request merged**